[DNM] drivers: clock_control: nrf: Update driver to use nrfx_clock and nrfx_power

anangl opened the pull request:
A draft PR based on https://github.com/zephyrproject-rtos/zephyr/pull/27422, to test if the approach proposed in https://github.com/zephyrproject-rtos/zephyr/pull/27422#issuecomment-672755626 is feasible and acceptable.

nordic-krch committed:
drivers: clock_control: nrf: Update driver to use nrfx_clock and nrfx_power

Updated driver implementation to use nrfx drivers.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: clock_control: nrf: Move USB part to the USB driver

Removed all nrf_power/USB related stuff from clock control
driver to USB driver.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

anangl committed:
soc: nordic: nrf51: Add missing peripheral Kconfig options

To indicate that the CLOCK and POWER peripherals are present in those
SoCs, thus the corresponding nrfx drivers can be used.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
cc @aescolar @carlescufi

anangl commented:
Rebased on master.

anangl commented:
Changed the nrf_hw_models module revision to the one from https://github.com/BabbleSim/ext_NRF52_hw_models/pull/12.

aescolar commented on boards/posix/nrf52_bsim/Kconfig.board:
```diff
@@ -14,6 +14,11 @@ config BOARD_NRF52_BSIM
+	# drivers for them can be used.
+	select HAS_HW_NRF_CLOCK
+	select HAS_HW_NRF_POWER
+	select HAS_NRFX
```
I wonder if there is a better way than this. If I understand correctly doing this, means both all the real nrfx and the HW models include paths are included. And things work only because of the include order(?).
If so, that is a bit fragile. It would be nicer if we could include selectively only the drivers folder, without pulling the HAL folders from both sides.
@anangl how would it sound to you, for example, to have separate HAS_NRFX_DRIVERS and has HAS_NRFX_HAL?

anangl replied:
> If I understand correctly doing this, means both all the real nrfx and the HW models include paths are included. And things work only because of the include order(?).
> If so, that is a bit fragile.

Yes, it relies on the include paths order. And the proper order is enforced by this change: https://github.com/zephyrproject-rtos/hal_nordic/pull/51/commits/3932f46487634d9edd475251507b5a21f4e0dc3f#diff-55df9be073e4151e19fc798a7c38f122. So I think it's not that fragile. And it gives you the possibility of replacing only the nrfx files that you need with their versions from ext_NRF52_hw_models and using others, if needed, directly.

> It would be nicer if we could include selectively only the drivers folder, without pulling the HAL folders from both sides.
> @anangl how would it sound to you, for example, to have separate HAS_NRFX_DRIVERS and has HAS_NRFX_HAL?

I don't really see a way of achieving this. There is no separate include path for nrfx HAL files defined, those are included as `<hal/nrf_*.h>` or `<nrfx/hal/nrf_*.h>` (for instance, in `radio.c` in the bluetooth subsystem).

anangl commented on boards/posix/nrf52_bsim/Kconfig.board:
```diff
@@ -14,6 +14,11 @@ config BOARD_NRF52_BSIM
+	# drivers for them can be used.
+	select HAS_HW_NRF_CLOCK
+	select HAS_HW_NRF_POWER
+	select HAS_NRFX
```
Yes, and it seems that `HAS_CMSIS_CORE` should not be selected by `HAS_NRFX`. It will be selected by `CPU_CORTEX_M` when the build targets the real SoC: https://github.com/zephyrproject-rtos/zephyr/blob/202ca01dcbe760b825f550df4d8411ad09f34df0/arch/arm/core/aarch32/Kconfig#L10
and that should be sufficient IMO.

I'll address it in a separate PR.

anangl committed:
tests/bluetooth/ll_settings: Prevent switch-unreachable warning

After the fallthrough comments have been replaced (see commit
741ad17a37807a5396e73c10cf3e5fb425bc87bb) with a macro expanding
to the corresponding attribute, those hints became statements,
and if one happens to occur before any switch case, a warning
reporting it as an unreachable statement is generated.
Prevent the above from happening for function ll_rx_dequeue() in
ull.c by enabling an option that will make some switch case to
appear before this problematic statement.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>
